### ISSUE TITLE
Bug: [no-floating-promises] Not erroring with multiple await statements in lexical scope

### ISSUE THREAD

**Issue by MarcCelani-at** [typescript-eslint/typescript-eslint#6416]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.9.3&sourceType=module&code=FDCGGcE8DsGMAIBmBXOAXAlge2krWAKASngG9h55Ydw15R4BeeABQCcsBbDcAUwDo2vAFa9YaYgG4QlUAHdQGOuy48BQ8FgA2AN14EALACYi0yhXho2kMhdkKl9M-AC+VUGlgALeAV4lyCxdpFyA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oDN4OBDfMwDmtYtA4BbSshTooiaBOiRwYAL4h1QA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
async function foo() {
  const a = Promise.reject();

  await Promise.resolve(42);
  
  try {
    await a;
  } catch (e) {

  };
}
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-floating-promises": "error"
  }
}
```

### Expected Result

This should error because there is a floating promise at the point the code awaits on a subset of promises created in this scope.

### Actual Result

There was no error

**Comment by MichaelMitchell-at** [comment]
This article illustrates why this is a bug
https://jakearchibald.com/2023/unhandled-rejections/

**Comment by bradzacher** [comment]
This rule doesn't enforce that you `await` all promises at once or in any specific order.
It simply looks at expression statements to determine if there is a floating promise.

In your code every single expression statement that is a promise is `await`ed, so the rule correctly does not flag anything.

Enforcing a specific way to `await` promises is going to be incorrect - there will be times you want to await one at a time, times you want to await them all at once, or times you want to await them sequentially. It's all about the context as to how you await them. The most important thing is that you await them all.

**Comment by MarcCelani-at** [comment]
Can we reopen this? From my perspective the goal of this rule is to avoid unhandled rejections. Maybe this is me understanding what a floating promise is, but if this is not the purpose of the rule, can we reopen this as a request for a new rule? The example I gave is overly simplistic in that i constructed fulfilled promises, but if the promises required a loop to evaluate, then it's possible for there to be an unhandled rejection that crashes the process with code that looks similar to this. That's the point, and why we need a lint rule to prevent this.

**Comment by MarcCelani-at** [comment]
"there will be times you want to await one at a time,"

If the one you do not await can be rejected, then your code has a bug.

**Comment by MichaelMitchell-at** [comment]
Here's a more illustrative example:
```ts
async function foo() {
  const a = Promise.reject(Error('handled too late'));

  await new Promise(resolve => setTimeout(resolve, 1000));
  
  try {
    await a;
  } catch (e) {

  };
}

foo();
```

Run this in `node` or in a Chrome console and observe the error

**Comment by bradzacher** [comment]
What you're debating here is whether or not every single promise should specifically be `await`ed so that its potential error can be appropriately handled.

EG you're saying this code is incorrect because `prom2` could reject, causing an unhandled rejection in the codebase
```ts
async function foo() {
  const prom1 = Promise.reject('oops');
  const prom2 = asyncJob();

  try {
    await prom1;
    await prom2; // this never executes
  } catch {}
}
```

Extrapolating, I guess you're saying the code should instead be
```ts
try { await prom1 } catch { }
try { await prom2 } catch { }

// or

await Promise.allSettled([prom1, prom2]);

// or some monstrously complicated thing like

try {
  await prom1;
} catch {  } finally {
  try {
    await prom2;
  } catch { ... }
}
```

In **_some_** cases - yes, it's correct that you should ensure that **_every_** single promise rejection is explicitly handled in some way. In other cases - just handling one rejection is more than enough. There simply isn't one hard-and-fast rule to run by.

For example imagine in a UI - you send two requests to the backend simultaneously. If both fail - in a lot of cases just showing the user one error is more than enough. Telling the user "Oops, something went wrong, sorry!" is fine - they probably don't care whether one or `N` errors occurred - to them it's the same - the process failed (again - maybe the user does care - it's not a hard-and-fast thing).

Under your rules using `Promise.all` or `Promise.race` are things that should never be used because they will exit at the first rejection and thus hide any subsequent rejections that might occur. In some cases that's completely correct - which is why `Promise.allSettled` was added to the API! Other times, no, it's not correct.

There's no hard-and-fast rule here as to how things should be done, it's contextually dependent on the workload and how the promises are constructed, and it's never going to be specifically one way or the other every single time.

Which is why a rule enforcing some specific style for multiple promises wouldn't make it into this project either - because for a lot of codebases it would be ripe with false positives.

**Comment by MichaelMitchell-at** [comment]
> What you're debating here is whether or not every single promise should specifically be awaited so that its potential error can be appropriately handled.

Not quite, we're saying that 

> EG you're saying this code is incorrect because prom2 could reject, causing an unhandled rejection in the codebase

Right

> Extrapolating, I guess you're saying the code should instead be

Those work, but this would also suffice too:
```ts
async function foo() {
  const prom1 = Promise.reject('oops');
  const prom2 = asyncJob()

  prom2.catch(() => {});  // Mark the rejection as handled

  try {
    await prom1;
    await prom2;
  } catch {}
}
```

This preserves the original intended semantics, with no downside other than slightly more obtuse code.

The article I linked to earlier suggests using a more readable utility function:
```ts
export function preventUnhandledRejections(...promises) {
  for (const promise of promises) promise.catch(() => {});
}
```

**Comment by MarcCelani-at** [comment]
For what it's worth, in our node repository, we *do* ask that developers avoid using Promise.all and Promise.race for the exact reason you mentioned.

But I understand that in some projects, you could just choose to allow an unhandled rejection happen. For that case, I agree that this rule as written is still useful in that it prevents a user from adding extra asynchronous code that they forget to await, but doesn't require them to prevent an unhandled rejection.

What do you think of a new rule then, no-unhandled-rejections, that attempts to detect and prevent unhandled rejections? This rule would disallow Promise.all, Promise.race, and the examples we showed. And ideally, there would be a way to express with the type system that a Promise cannot be rejected so as to make this rule easier to use (it is okay to use Promise.all, Promise.race, or write code like the above if you know that the promise cannot be rejected).

**Comment by bradzacher** [comment]
The issue is that there is not a standard way to represent an "unthrowable" promise in the type-system.

So if you define a helper function like `preventUnhandledRejections`, then a rule will continue to report on those promises. If you reassign a promise or pass it through a function then a rule can't detect the change and will report on those cases.
Which makes the rule dead in the water in a general purpose plugin like ours - we'd just create noise.

If TS had tools to represent this, then we could in turn detect it and build a rule around it.

**Comment by MarcCelani-at** [comment]
@bradzacher : +1, I also realized this as well. I've been working on a separate rule proposal for the past hour (#6418) and I actually pointed out that something like an UnrejectablePromise would be super helpful for this kind of rule.

I'd humbly ask that we keep it open and I can open an issue in microsoft/TypeScript for us to discuss such a type.

**Comment by MichaelMitchell-at** [comment]
> So if you define a helper function like preventUnhandledRejections, then a rule will continue to report on those promises. If you reassign a promise or pass it through a function then a rule can't detect the change and will report on those cases.

That's a good point and something to consider for #6418 would be to add a config setting to provide the names of functions that handle any promises passed as arguments, e.g.
```ts
{
  "rules": {
    // ...
    "@typescript-eslint/no-unhandled-promise-rejections": ["warn", {
      "promiseHandlerFunctionNames": ["preventUnhandledRejections"],
    }]
  }
}
```

**Comment by MarcCelani-at** [comment]
Let's move the discussion to #6418 please.
